README.md
# Kitchen::Lxd

[author]:           https://github.com/DracoAter
[issues]:           https://github.com/DracoAter/kitchen-lxd/issues
[license]:          https://github.com/DracoAter/kitchen-lxd/blob/master/LICENSE
[repo]:             https://github.com/DracoAter/kitchen-lxd
[hostnamectl bug]: https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/1575779
[lxd_doc]: https://github.com/lxc/lxd/blob/master/doc/containers.md

[![Build Status](https://travis-ci.org/DracoAter/kitchen-lxd.svg?branch=master)](https://travis-ci.org/DracoAter/kitchen-lxd)
[![Gem Version](https://badge.fury.io/rb/kitchen-lxd.svg)](https://badge.fury.io/rb/kitchen-lxd)

- [Requirements](#requirements)
	- [Lxd](#lxd)
- [Installation and Setup](#installation-and-setup)
- [Configuration](#configuration)
	- [Driver](#driver)
	- [Transport](#transport)
- [Development](#development)
- [Authors](#authors)
- [License](#license)

A Test Kitchen Driver (with Transport) for Lxd.

## Requirements

### Lxd

Lxd version of 2.3 (the one where "lxc network" commands were introduced) or higher is required
for this driver which means that a native package must be installed on the system running Test
Kitchen.

You do not have to prepare any container image specifically, like downloading it or installing ssh
server. The driver will download container image automatically from the provided remote server,
if it's not available locally. Also you can use `lxd` transport instead of default `ssh`. Which
means files will be uploaded to container using `lxc file push` command.

## Installation and Setup

Install using command line:

```bash
gem install kitchen-lxd
```

## Configuration

Example config file may look like this:

```yaml
---
driver:
  name: lxd
  binary: lxc # this is default
  remote: images # default
  network: lxdbr0 # default
  fix_chef_install: false # default
  fix_hostnamectl_bug: true # default
  config: {} # default

transport:
  name: lxd
```

Default values can be omitted, so the minimal config file looks like this:

```yaml
---
driver:
  name: lxd

transport:
  name: lxd
```

You can specify different keys/values under `config`. Look for list of the available items in
[Lxd Documentation][lxd_doc]. For example, if you need to start container in privileged mode, add:

```yaml
---
driver:
  name:lxd
  config:
    security.privileged: true
```

### Driver

Available options:

Name | Description | Type | Default
-----|-------------|------|--------
binary | Path to lxc executable. | String | `lxc`
remote | Remote LXD server to download image from, if it does not exist locally. | String | `images`
network | Network bridge to attach to container. | String | `lxdbr0`
wait_until_ready | Wait for the network to come up. | Boolean | `true`
config | Config key/value to apply to the new container. This hash is passed to `lxc init` command as --config parameters. | Hash | `{}`
fix_chef_install | Apply fix, to make available installation of Chef Omnibus package. | Boolean | `false`
fix_hostnamectl_bug | Apply workaround to Ubuntu [hostnamectl bug][hostnamectl bug] in LXD. | Boolean | `true`

## Development

- Source hosted at [GitHub][repo]
- Report issues/questions/feature requests on [GitHub Issues][issues]

Pull requests are very welcome! Make sure your patches are well tested.
Ideally create a topic branch for every separate change you make. For
example:

1. Fork the repo
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Authors

Created and maintained by [Juri Timošin][author].

## License

Apache 2.0 (see [LICENSE][license])


CHANGELOG.md
## 0.3.0 (June 2, 2018)

- Introduce `config` parameter in configuration that is passed to `lxc init` as `--config`
  parameter.

## 0.2.2 (January 30, 2018)

- Fix upload bug.

## 0.2.1 (October 13, 2017)

- Add workaround to Ubuntu hostnamectl bug in LXD (https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/1575779)

## 0.2.0 (October 11, 2017)

- Add LXD transport.
- Remove ssh transport support.
- Allow specifying network bridge to attach to.
- Allow specifying path to lxc executable.
- Support downloading image from remote, if it's not available locally.

## 0.1.1 (February 08, 2017)

- Support Lxd version 2.8.

## 0.1.0 (February 07, 2017)

- Initial release.


test/unit/test_helper.rb

# frozen_string_literal: true

#
# Author:: Juri Timoshin (<draco.ater@gmail.com>)
#
# Copyright (C) 2017, Juri Timoshin
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'simplecov'
require 'minitest/autorun'

require_relative '../../lib/kitchen/driver/lxd'


test/integration/test_helper.rb
# -*- encoding: utf-8 -*-
#
# Author:: Juri Timošin (<draco.ater@gmail.com>)
#
# Copyright (C) 2017, Juri Timošin
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'simplecov'
require 'minitest/autorun'

require_relative '../../lib/kitchen/driver/lxd'


test/unit/kitchen_driver_lxd_container_test.rb

# frozen_string_literal: true

#
# Author:: Juri Timoshin (<draco.ater@gmail.com>)
#
# Copyright (C) 2017, Juri Timoshin
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require_relative 'test_helper'
require 'logger'

module Kitchen
	module Driver
		class Lxd
			module UnitTest
				class ContainerTest < Minitest::Test
					def setup
						@subj = Lxd::Container.new(::Logger.new(StringIO.new), container: 'kitchen-lxd-test',
							image: 'alpine/3.6', binary: 'lxc', remote: 'images')
					end

					def test_constructor
						assert_equal 'kitchen-lxd-test', @subj.instance_variable_get(:@name)
						assert_equal 'alpine/3.6', @subj.instance_variable_get(:@image)
						assert_equal 'lxc', @subj.instance_variable_get(:@binary)
						assert_equal 'images', @subj.instance_variable_get(:@remote)
					end

					def test_login_command
						assert_equal 'lxc exec kitchen-lxd-test -- $(lxc exec kitchen-lxd-test -- head -1 '\
							'/etc/passwd | cut -d: -f7)', @subj.login_command.command
					end
				end
			end
		end
	end
end


lib/kitchen/driver/lxd/container.rb

# frozen_string_literal: true

#
# Author:: Juri Timoshin (<draco.ater@gmail.com>)
#
# Copyright (C) 2017, Juri Timoshin
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'kitchen'
require 'json'

module Kitchen
	module Driver
		class Lxd < Kitchen::Driver::Base
			class Container
				include ShellOut
				include Logging

				attr_reader :logger
				attr_reader :state

				def initialize(logger, opts)
					@logger = logger
					@name = opts[:container]
					@image = opts[:image]
					@remote = opts[:remote]
					@binary = opts[:binary]
					@fix_hostnamectl_bug = opts[:fix_hostnamectl_bug]
				end

				def init(config={})
					return if created?
					download_image unless image_exists?
					config_args = config.map{|k, v| "-c #{k}='#{v}'" }.join(' ')
					run_command "#{@binary} init #{@image} #{@name} #{config_args}"
				end

				def attach_network(network)
					return if device_attached? network
					run_command "#{@binary} network attach #{network} #{@name}"
				end

				def start
					return if running?
					run_command "#{@binary} start #{@name}"
				end

				def destroy
					return unless created?
					run_command "#{@binary} delete #{@name} --force"
				end

				def wait_until_ready
					info 'Wait for network to become ready.'
					9.times do
						s = fetch_state[:state].nil? ? @state[:State] : @state[:state]
						inet = s.dig(:network, :eth0, :addresses)&.detect do |i|
							i[:family] == 'inet'
						end
						return inet[:address] if inet
						sleep 1 unless defined?(Kitchen::Driver::UnitTest)
					end
					nil
				end

				def download_image
					run_command "#{@binary} image copy --copy-aliases #{@remote}:#{@image} local:"
				end

				def execute(command)
					return if command.nil? or command.empty?
					fix_hostnamectl_bug if @fix_hostnamectl_bug
					run_command "#{@binary} exec #{@name} -- #{command}"
				end

				def login_command
					LoginCommand.new("#{@binary} exec #{@name} -- $(#{@binary} exec #{@name} -- head -1 "\
						'/etc/passwd | cut -d: -f7)', {})
				end

				def upload(locals, remote)
					return if locals.nil? or locals.empty?
					run_command "#{@binary} file push -r #{locals.join(' ')} #{@name}#{remote}"
				end

				def fix_chef_install(platform)
					case platform
					when /ubuntu/, /debian/
						execute 'apt-get update'
						execute 'apt-get install -y wget'
					when /rhel/, /centos/
						execute 'yum install -y sudo wget'
					end
				end

				def fix_hostnamectl_bug
					logger.info 'Replace /usr/bin/hostnamectl with /usr/bin/true, because of bug in Ubuntu'\
						'. (https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/1575779)'
					run_command "#{@binary} exec #{@name} -- ln -fs /usr/bin/true /usr/bin/hostnamectl"
				end

    #				private

				def image_exists?
					!JSON.parse(
						run_command("#{@binary} image list #{@image} --format json"), symbolize_names: true
					).empty?
				end

				def fetch_state
					@state = JSON.parse(
						run_command("#{@binary} list #{@name} --format json"), symbolize_names: true
					).first
				end

				def running?
					fetch_state[:status] == 'Running'
				end

				def created?
					!fetch_state.nil?
				end

				def device_attached?(network)
					fetch_state.dig(:devices, network.to_sym) ? true : false
				end
			end
		end
	end
end


lib/kitchen/driver/lxd.rb

# frozen_string_literal: true

#
# Author:: Juri Timoshin (<draco.ater@gmail.com>)
#
# Copyright (C) 2017, Juri Timoshin
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'kitchen'
require_relative 'lxd/container'

module Kitchen
	module Driver
		# Lxd driver for Kitchen.
		#
		# @author Juri Timoshin <draco.ater@gmail.com>
		class Lxd < Kitchen::Driver::Base
			include ShellOut
			MIN_LXD_VERSION = '2.3'

			kitchen_driver_api_version 2

			default_config :binary, 'lxc'
			default_config :remote, 'images'
			default_config :network, 'lxdbr0'
			default_config :config, {}
			default_config :wait_until_ready, true
			default_config :fix_chef_install, false
			default_config :fix_hostnamectl_bug, true # https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/1575779

			default_config :image do |driver|
				driver.instance.platform.name
			end

			default_config :container do |driver|
				driver.instance.name
			end

			def create(state)
				container.init(config[:config])
				container.attach_network config[:network] if config[:network]
				container.start

				if config[:wait_until_ready]
					state[:hostname] = instance.transport.connection(state).wait_until_ready
				end
				container.fix_chef_install(instance.platform.name) if config[:fix_chef_install]
				container.fix_hostnamectl_bug if config[:fix_hostnamectl_bug]
			end

			def destroy(state)
				instance.transport.connection(state).close
				state.delete :hostname
				container.destroy
			end

			def verify_dependencies
				version = run_command("#{config[:binary]} --version").strip
				return if Gem::Version.new(version) >= Gem::Version.new(MIN_LXD_VERSION)
				raise UserError, "Detected old version of Lxd (#{version}), please upgrade to version "\
					"#{MIN_LXD_VERSION} or higher."
			end

			def container
				@container ||= Lxd::Container.new(logger, config)
			end
		end
	end
end
